Removal of repeatable parameters

Eclipse-Dominator opened the pull request:
This PR an alternative that Fixes #162 (other alternatives: #176, #181)

In this PR, we treat repeated parameters as an error.

```
If a single-valued parameter is repeated multiple times in a 
command, only the last occurrence is taken.  This allows the user
to correct a mistake in a parameter value by simply typing it again,
which may be faster than moving the cursor back to the original
parameter values and editing it. This feature serves as a good
example of 'going the extra mile to optimize the application for
target users' (i.e., users who can type fast).
However, this feature can cause certain user errors to go unnoticed
e.g., unintended repetition caused by a typo in parameter name

Alternatives considered:
(a) Retain current behavior but give a friendly warning to the user
    that a parameter was 'overridden'. 
(b) Remove this feature i.e., treat repeated single-valued
    parameters    as an error, and reject the command with an
    error message.

Alternative (a) requires significantly more complicated code
(see #181), steepening the learning curve for for new programmers.
Given that this application is primarily meant for training
novice programmers, the impact on the learning curve is not worth
the benefit of retaining this feature.

Therefore, let's do option (b).
```

damithc commented:
@Eclipse-Dominator Now that we have three alternatives, what your thoughts about selecting one?
Also, in each of the three PR descriptions, you can mention the other two so that anyone visiting the PR can see how they fit into the big picture.

damithc commented:
> Repeated Parameters show warning: #176
> Repeated Parameters show warning of repeated prefixes: #181
> Repeated Parametes treated as errors: #190

Updated the [PR description](https://github.com/se-edu/addressbook-level3/pull/190#issue-1744773620) instead.

damithc commented on src/main/java/seedu/address/logic/parser/EditCommandParser.java:
```diff
@@ -78,5 +87,4 @@ private Optional<Set<Tag>> parseTagsForEdit(Collection<String> tags) throws Pars
         Collection<String> tagSet = tags.size() == 1 && tags.contains("") ? Collections.emptySet() : tags;
         return Optional.of(ParserUtil.parseTags(tagSet));
     }
-
```
Avoid unrelated changes

damithc commented on src/main/java/seedu/address/logic/parser/AddCommandParser.java:
```diff
@@ -44,6 +45,13 @@ public AddCommand parse(String args) throws ParseException {
+
+        if (!duplicatePrefixes.isEmpty()) {
+            throw new ParseException(Messages.formatDuplicatePrefixesToMessage(duplicatePrefixes));
+        }
```
Instead of the above, would it be better to be able do this?
```java
argMultimap.verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS);
```

damithc commented on src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:
```diff
@@ -51,27 +58,59 @@ public void parse_allFieldsPresent_success() {
-        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_AMY + EMAIL_DESC_BOB
-                + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
+    @Test
+    public void parse_repeatedNonTagValue_success() {
```
`_success()` or `_exceptionThrown()`?

damithc reviewed (commented):
Some initial comments ...

damithc commented:
@Eclipse-Dominator As discussed, let's proceed with this approach.

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,22 @@ public class Messages {
+
+        Set<String> duplicateFields = duplicatePrefixes.stream().map(Prefix::toString).collect(Collectors.toSet());
+
+
```
Is double blank lines necessary?

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,22 @@ public class Messages {
+    /**
+     * Formats the duplicate prefixes into an error message.
+     */
+    public static String formatDuplicatePrefixesToMessage(Set<Prefix> duplicatePrefixes) {
```
`getErrorMessageForDuplicatePrefixes`?

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +59,19 @@ public List<String> getAllValues(Prefix prefix) {
+     * the argument multimap.
+     */
+    public Set<Prefix> getDuplicatePrefixes(Prefix... prefixes) {
+        Set<Prefix> ret = new HashSet<>();
```
`ret`: use full words unless the short form is widely used.

damithc commented on src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:
```diff
@@ -51,27 +58,59 @@ public void parse_allFieldsPresent_success() {
-                + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
+    @Test
+    public void parse_repeatedNonTagValue_success() {
+        String validExpectedPerson = NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
```
variable name suggests this is a person object, but it is not.

damithc commented on src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:
```diff
@@ -51,27 +58,59 @@ public void parse_allFieldsPresent_success() {
-        // multiple addresses - last address accepted
-        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_AMY
-                + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
+        // multiple fields repeated
```
It's more typical to test for individual errors before testing for a mix of errors.

damithc reviewed (changes requested):
Added a few more comments.

jundatan commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,21 @@ public class Messages {
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The person index provided is invalid";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d persons listed!";
+    public static final String MESSAGE_DUPLICATE_FIELDS = "Detected multiple values for the following field%s: "
+            + "%s.\nPlease only a single value for those fields\n";
```
I feel that this Message should be worded properly.  Perhaps it will be better to have a message like "Multiple values detected for the following field%s: " + "%s.\nOnly one value is allowed for these fields\n".

jundatan reviewed (commented):
Suggestion to change the message for duplicate person.

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,21 @@ public class Messages {
+
+        return String.format(MESSAGE_DUPLICATE_FIELDS, duplicateFields.size() > 1 ? "s" : "",
+                String.join(" ", duplicateFields), duplicateFields.size() > 1 ? "these" : "this");
+    }
```
I think we are complicating this code too much for too little payback. No need to bother with singular/plural differences. Something like the following?

`Multiple values specified for the following single-valued filed(s): ...`

Eclipse-Dominator replied:
Personally, I think over complicating this code is okay here since everything related to the messages have been abstracted to within `Messages`. So, there won't be any additional code duplication and its self-contained to within `Messages`.

That being said, I think changing it the message to
```
Multiple values specified for the following single-valued field(s): ...
Please only supply a single value for the field(s) mentioned above.
```
can make it appear more user-friendly?

damithc replied:
>Personally, I think over complicating this code is okay here since everything related to the messages have been abstracted to within Messages. So, there won't be any additional code duplication and its self-contained to within Messages.

**Always** be as economical as possible with code. Every line of code is an additional burden every project member has to carry for the rest of the lifetime of the project.

>That being said, I think changing it the message to
>
>Multiple values specified for the following single-valued field(s): ...
Please only supply a single value for the field(s) mentioned above.
can make it appear more user-friendly?

Not totally opposed, but I have some concerns:
1. Feels a bit redundant. Check how the UI present this error to the user. Perhaps the first sentence is enough for the user to understand the problem.
2. Check other error messages. If they don't have `Please do such and such ...` part, we don't want to make this specific error message different from others.

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +62,24 @@ public List<String> getAllValues(Prefix prefix) {
     }
+
+    /**
+     * throws a ParseException if any of the prefixes given in {@code prefixes} appeared more than
```
```suggestion
     * Throws a (@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
```

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +62,24 @@ public List<String> getAllValues(Prefix prefix) {
+
+    /**
+     * throws a ParseException if any of the prefixes given in {@code prefixes} appeared more than
+     * once in the argument multimap.
```
```suggestion
     * once among the arguments.
```

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +62,24 @@ public List<String> getAllValues(Prefix prefix) {
+            return;
+        }
+
+        throw new ParseException(Messages.getDuplicatePrefixesToMessage(duplicatedPrefixes));
```
alternative:
```suggestion
        if (!duplicatedPrefixes.isEmpty()) {
            throw new ParseException(Messages.getDuplicatePrefixesToMessage(duplicatedPrefixes));
        }
        
```

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,21 @@ public class Messages {
+    /**
+     * Formats the duplicate prefixes into an error message.
+     */
+    public static String getDuplicatePrefixesToMessage(Set<Prefix> duplicatePrefixes) {
```
alternative to consider:
```suggestion
    public static String getDuplicatePrefixesToMessage(Prefix... duplicatePrefixes) {
```

Eclipse-Dominator replied:
I kept it intentionally this to avoid ordering of the fields in the message to avoid complications when doing unit test etc. 
It is possible to sort the field within this method but I feel that supplying it with a `Set` will make it cleaner.

damithc replied:
Instead of creating `Set` objects everywhere this method is called, you can create the `Set` inside the method at the start?

Eclipse-Dominator replied:
Yes, but the code technically is really only called once in production code and the rest of it is called in unit test.

Which is called below.
```java
    /**
     * Throws a {@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
     * once among the arguments.
     */
    public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseException {
        Set<Prefix> duplicatedPrefixes = new HashSet<>();
        Set<Prefix> prefixSet = Set.of(prefixes);
        argMultimap.forEach((key, list) -> {
            if (list.size() > 1 && prefixSet.contains(key)) {
                duplicatedPrefixes.add(key);
            }
        });

        if (!duplicatedPrefixes.isEmpty()) {
            throw new ParseException(Messages.getDuplicatePrefixesToMessage(duplicatedPrefixes));
        }
    }
```
Converting `getDuplicatePrefixesToMessage()` to take in `Prefix...` or `Prefix[]` will involve calling `duplicatedPrefixes.toArray(Prefix[]::new))` which will cause it to be converted once again into a `Set` object when inside the method. 

This is true unless we are using a fixed-size array for `duplicatedPrefixes` or alternative we can convert the whole process as a java stream and convert it to an array. 

I will experiment processing the above with java stream and see its readability in the next version.

Eclipse-Dominator replied:
I think the code is cleaner with java stream, I will update the relevant code.

damithc replied:
BTW, you are doing something very similar in here:
```java
 public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseException {
        Set<Prefix> duplicatedPrefixes = new HashSet<>();
        Set<Prefix> prefixSet = Set.of(prefixes);
```

damithc reviewed (commented):
Some suggestions added

damithc commented on docs/UserGuide.md:
```diff
@@ -57,8 +57,8 @@ AddressBook Level 3 (AB3) is a **desktop app for managing contacts, optimized fo
-* If a parameter is expected only once in the command but you specified it multiple times, only the last occurrence of the parameter will be taken.<br>
-  e.g. if you specify `p/12341234 p/56785678`, only `p/56785678` will be taken.
+* Parameters that is expected to only appear once in the command will be treated as an invalid command if you specified it multiple times.<br>
+  e.g. if you specify `p/12341234 p/56785678`, the application will treat this as an invalid command.
```
Good catch.
1. As this was a 'special' behavior, we had to specifically mention it in the UG. But now that we have removed that special behavior, just removing the said documentation is enough. No need to explain the updated behavior, as that is what one would expect anyway.
2. This UG change is tightly related to the behavior change. Should be in the same commit. Each commit should take the code from one 'correct' state to another 'correct' state. A commit that updates the behavior without updating the UG takes the code to an incorrect state.

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +14,21 @@ public class Messages {
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d persons listed!";
+    public static final String MESSAGE_DUPLICATE_FIELDS =
+            "Multiple values specified for the following single-valued field%1$s: %2$s\n"
+                    + "Please only supply a single value for the field%1$s mentioned above.";
```
Line wrapping is odd.

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +166,37 @@ public void parse_oneFieldSpecified_success() {
 
     @Test
-    public void parse_multipleRepeatedFields_acceptsLast() {
+    public void parse_multipleRepeatedFields_failure() {
```
I think I'm the one who suggested to use `_failure()` but I regret that now. We don't expect tests to fail. A better way to indicated the expected results is `_exceptionThrown()`

Eclipse-Dominator replied:
Perhaps we can change `_failure()` to `_exceptionThrown` or some more appropriate terms to standardize the unit test formatting in a separate PR for the other unit tests

damithc replied:
Ah... so it is used elsewhere as well. In that case leave it be, for consistency.

damithc reviewed (changes requested):
Some further suggestions

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +15,19 @@ public class Messages {
+    public static String getDuplicatePrefixesToMessage(Prefix... duplicatePrefixes) {
+        if (duplicatePrefixes.length == 0) {
+            return "";
+        }
```
This should not even happen, right?
In that case, use an assert instead.

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +15,19 @@ public class Messages {
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The person index provided is invalid";
     public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d persons listed!";
+    public static final String MESSAGE_DUPLICATE_FIELDS =
+            "Multiple values specified for the following single-valued field(s): %s";
```
There should be a blank line here?

damithc replied:
Do we need the `%s`?

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -9,5 +15,19 @@ public class Messages {
+        }
+
+        Set<String> duplicateFields =
+                Stream.of(duplicatePrefixes).map(Prefix::toString).collect(Collectors.toSet());
```
Our policy is to line-wrap only when close to the line limit, right? e.g.,
```suggestion
        Set<String> duplicateFields = Stream.of(duplicatePrefixes).map(Prefix::toString)
                .collect(Collectors.toSet());
```

Follow that policy consistently.

damithc commented on src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:
```diff
@@ -51,27 +56,76 @@ public void parse_allFieldsPresent_success() {
+        assertParseFailure(parser,
+                validExpectedPersonString + PHONE_DESC_AMY + EMAIL_DESC_AMY + NAME_DESC_AMY
+                        + ADDRESS_DESC_AMY + validExpectedPersonString,
+                Messages.getDuplicatePrefixesToMessage(PREFIX_NAME, PREFIX_ADDRESS, PREFIX_EMAIL, PREFIX_PHONE));
```
This can be moved just above `// invalid value followed by valid value`. Otherwise it appears as if belonging to `// valid value followed by invalid value` category.

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -90,10 +92,6 @@ public void parse_invalidValue_failure() {
-        // valid phone followed by invalid phone. The test case for invalid phone followed by valid phone
-        // is tested at {@code parse_invalidValueFollowedByValidValue_success()}
-        assertParseFailure(parser, "1" + PHONE_DESC_BOB + INVALID_PHONE_DESC, Phone.MESSAGE_CONSTRAINTS);
-
```
Good catch.

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +164,34 @@ public void parse_oneFieldSpecified_success() {
-        expectedCommand = new EditCommand(targetIndex, descriptor);
-        assertParseSuccess(parser, userInput, expectedCommand);
+        assertParseFailure(parser, userInput,
+                Messages.getDuplicatePrefixesToMessage(PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS));
```
It seems you do more extensive testing of this aspect for the Add command than the Edit command. Perhaps we should make a note of that somewhere in this class? e.g.,
`/* More extensive testing of duplicate parameter detections is done in ...*/`

damithc reviewed (changes requested):
Some suggestions added (mostly cosmetic)

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +164,39 @@ public void parse_oneFieldSpecified_success() {
+        String userInput = targetIndex.getOneBased() + INVALID_PHONE_DESC + PHONE_DESC_BOB;
+
+        assertParseFailure(parser, userInput,
+                        Messages.getDuplicatePrefixesToMessage(PREFIX_PHONE));
```
Unnecessary line break?

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +164,39 @@ public void parse_oneFieldSpecified_success() {
+        userInput = targetIndex.getOneBased() + PHONE_DESC_BOB + INVALID_PHONE_DESC;
+
+        assertParseFailure(parser, userInput,
+                        Messages.getDuplicatePrefixesToMessage(PREFIX_PHONE));
```
Same here

damithc commented on docs/UserGuide.md:
```diff
@@ -57,9 +57,6 @@ AddressBook Level 3 (AB3) is a **desktop app for managing contacts, optimized fo
 * Parameters can be in any order.<br>
```
Added a few cosmetic suggestions.

Commit message:
```
Currently, the behavior of repeated parameters in AddCommand and
EditCommand is to consider the last value used. This behavior can be
confusing and difficult to work with. Therefore, instead of allowing
repeatable parameters, it is better to present them as a Command Error
to the user.
```
Has vague parts e.g., 'difficult to work with' (why is it difficult to work with?)
You can refine this based on the description given in https://github.com/se-edu/addressbook-level3/issues/162


```
To implement this change, the checking for duplicate parameters should
be performed before checking the validity of individual fields. In other
words, if an Add or Edit Command contains multiple duplicate fields that
are all invalid, the parsing error for duplicate fields will take precedence
over other checks for individual parameters.
```
* What is the justification for choosing this behavior? Commit messages should cover the WHY aspect. 
* This explanation doesn't cover cases of multiple values only _some_ of which are invalid. It only mentions cases where _all_ of them are invalid.

damithc replied:
I added these comments sometime back but it was stuck in 'draft' status.

Eclipse-Dominator commented:
Resolve rebase conflicts.

damithc commented:
@Eclipse-Dominator Looks like the rebase wasn't entirely correct. There are some unrelated code in this PR.

damithc commented on docs/UserGuide.md:
```diff
@@ -57,9 +57,6 @@ AddressBook Level 3 (AB3) is a **desktop app for managing contacts, optimized fo
 * Parameters can be in any order.<br>
   e.g. if the command specifies `n/NAME p/PHONE_NUMBER`, `p/PHONE_NUMBER n/NAME` is also acceptable.
```
Suggested tweaks to the commit message:

```
Remove support for repeated parameters in commands

If a single-valued parameter is repeated multiple times in a 
command, only the last occurrence is taken.  This allows the user
to correct a mistake in a parameter value by simply typing it again,
which may be faster than moving the cursor back to the original
parameter values and editing it. This feature serves as a good
example of 'going the extra mile to optimize the application for
target users' (i.e., users who can type fast).
However, this feature can cause certain user errors to go unnoticed
e.g., unintended repetition caused by a typo in parameter name

Alternatives considered:
(a) Retain current behavior but give a friendly warning to the user
    that a parameter was 'overridden'. 
(b) Remove this feature i.e., treat repeated single-valued
    parameters    as an error, and reject the command with an
    error message.

Alternative (a) requires significantly more complicated code
(see #181), steepening the learning curve for for new programmers.
Given that this application is primarily meant for training
novice programmers, the impact on the learning curve is not worth
the benefit of retaining this feature.

Therefore, let's do option (b).
```

Changes: 
1. Use imperative mood in subject
2. Added more details about the alternatives and the decision

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -11,6 +16,20 @@ public class Messages {
+    /**
+     * Formats the duplicate prefixes into an error message.
+     */
+    public static String getDuplicatePrefixesToMessage(Prefix... duplicatePrefixes) {
```
```suggestion
    public static String getErrorMessageForDuplicatePrefixes(Prefix... duplicatePrefixes) {
```

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +61,18 @@ public List<String> getAllValues(Prefix prefix) {
+    public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseException {
+        Prefix[] duplicatedPrefixes = Stream.of(prefixes).distinct().filter(
+                prefix -> argMultimap.containsKey(prefix) && argMultimap.get(prefix).size() > 1)
+                .toArray(Prefix[]::new);
```
Is this indentation correct?  Alternative:
```suggestion
        Prefix[] duplicatedPrefixes = Stream.of(prefixes).distinct()
                .filter(prefix -> argMultimap.containsKey(prefix) && argMultimap.get(prefix).size() > 1)
                .toArray(Prefix[]::new);
```

damithc commented on src/main/java/seedu/address/commons/core/Messages.java:
```diff
@@ -1,5 +1,10 @@
+import java.util.stream.Stream;
+
+import seedu.address.logic.parser.Prefix;
 import seedu.address.model.person.Person;
```
Dependency on `model` is already a problem, and we are supposed to fix that separately. Now we are adding a dependency on `logic`. Would that be a further problem?

damithc replied:
After another look, I think this should be fine, as we plan to move Messages to the `logic` package.

damithc reviewed (changes requested):
Some comments added (did not review the test code though).

damithc commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +164,37 @@ public void parse_oneFieldSpecified_success() {
+        /*
+         * More extensive testing of duplicate parameter detections is done in
+         * AddCommandParserTest#parse_repeatedNonTagValue_failure()
+         */
```
Should we use `//` here?

damithc reviewed (commented):
Just a minor question. I think we are close to home on this one. Let me ask others to have a look as well.

damithc commented:
@se-edu/tech-team-level1 for your review ...

dlimyy commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +61,18 @@ public List<String> getAllValues(Prefix prefix) {
     }
+
+    /**
+     * Throws a {@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
```
For the javadocs, I believe the `@param` and `@throws` tags have to be added
```suggestion
    /**
     * Throws a {@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
     * once among the arguments.
     *
     * @param prefixes List of prefixes supplied by the user
     * @throws ParseException If duplicate prefixes are present
     */
```

damithc replied:
@dlimyy Good catch that some tags are missing. Our policy is javadoc tags can be omitted if not adding value e.g., repeating the same information already available elsewhere in the javadoc

dlimyy commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +61,18 @@ public List<String> getAllValues(Prefix prefix) {
+     * Throws a {@code ParseException} if any of the prefixes given in {@code prefixes} appeared more than
+     * once among the arguments.
+     */
+    public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseException {
```
Could consider shortening the name of the method to verifyNoDuplicatePrefixes as the 'For' does not seem to be very necessary

damithc replied:
It reads as a natural statement e.g., `verifyNoDuplicatePrefixesFor(PREFIX_NAME, PREFIX_ADDRESS)` i.e., verify no duplicate prefixes for A, B, ..
But OK to remove. 
Can also consider `verifyNoDuplicatesFor` as Prefix is implied by the parameter type.

dlimyy commented on src/test/java/seedu/address/logic/parser/EditCommandParserTest.java:
```diff
@@ -166,36 +164,37 @@ public void parse_oneFieldSpecified_success() {
+
+        assertParseFailure(parser, userInput, Messages.getErrorMessageForDuplicatePrefixes(PREFIX_PHONE));
+
+        // mulltiple valid field repeated
```
Some minor grammatical errors here. It should be `multiple valid fields repeated` instead.

dlimyy reviewed (commented):
Overall, LGTM! Just some minor suggestions below

SPWwj commented on docs/UserGuide.md:
```diff
@@ -57,9 +57,6 @@ AddressBook Level 3 (AB3) is a **desktop app for managing contacts, optimized fo
-* If a parameter is expected only once in the command but you specified it multiple times, only the last occurrence of the parameter will be taken.<br>
-  e.g. if you specify `p/12341234 p/56785678`, only `p/56785678` will be taken.
-
 * Extraneous parameters for commands that do not take in parameters (such as `help`, `list`, `exit` and `clear`) will be ignored.<br>
   e.g. if the command specifies `help 123`, it will be interpreted as `help`.
```
```suggestion
* Extra parameters will be ignored.
  e.g. if the command is help 123`, it will be interpreted as help. This is also the case for `list`, `exit`, and `clear`.
```

damithc replied:
Not related to this PR?

SPWwj commented on src/main/java/seedu/address/logic/Messages.java:
```diff
@@ -11,6 +16,20 @@ public class Messages {
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
     public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The person index provided is invalid";
```
```suggestion
    public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "Invalid person index";
```

Eclipse-Dominator replied:
I think this is not related to the current PR, I think we can fine tune this messages in another PR.

SPWwj commented on src/main/java/seedu/address/logic/Messages.java:
```diff
@@ -11,6 +16,20 @@ public class Messages {
+                "Multiple values specified for the following single-valued field(s): ";
+
+    /**
+     * Formats the duplicate prefixes into an error message.
```
```suggestion
     * Generates a formatted error message indicating the duplicate prefixes.
```

damithc replied:
`a formatted error message` -> `an error message`?

damithc commented on src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:
```diff
@@ -57,4 +61,22 @@ public List<String> getAllValues(Prefix prefix) {
+     *
+     * @param prefixes Prefixes to check for duplicates
+     * @throws ParseException If any of the prefixes given in {@code prefixes} appeared more than once among the
+     *                       arguments
```
No need, as I mentioned in my comment.

damithc commented on src/main/java/seedu/address/logic/Messages.java:
```diff
@@ -11,6 +16,20 @@ public class Messages {
+                "Multiple values specified for the following single-valued field(s): ";
+
+    /**
+     * Generates a formatted error message indicating the duplicate prefixes.
```
`a formatted error message` -> `an error message` (as I mentioned in my previous comment)
While you are at it, `Returns`  may be better than `Generates`.

damithc reviewed (changes requested):
Two commented you missed.

Eclipse-Dominator committed:
Remove support for repeated parameters in commands

If a single-valued parameter is repeated multiple times in a command,
only the last occurrence is taken. This allows the user to correct a
mistake in a parameter value by simply typing it again, which may be
faster than moving the cursor back to the original parameter values and
editing it. This feature serves as a good example of 'going the extra
mile to optimize the application for target users' (i.e., users who can
type fast).
However, this feature can cause certain user errors to go unnoticed
e.g., unintended repetition caused by a typo in parameter name

Alternatives considered:
(a) Retain current behavior but give a friendly warning to the user
    that a parameter was 'overridden'.
(b) Remove this feature i.e., treat repeated single-valued parameters
    as an error, and reject the command with an error message.

Alternative (a) requires significantly more complicated code (see #181),
steepening the learning curve for for new programmers. Given that this
application is primarily meant for training novice programmers, the
impact on the learning curve is not worth the benefit of retaining this
feature.

Therefore, let's do option (b).

SPWwj reviewed (approved):
Code works as expected. LGTM.
